File: Algo&Comp/cyclic_dfs.cpp
#include <iostream>
#include <vector>

using namespace std;

bool dfsCycle(int node, vector<vector<int>>& graph, vector<bool>& visited, vector<bool>& recStack, vector<int>& path) {
    visited[node] = true;
    recStack[node] = true;
    path.push_back(node);

    for (int neighbor = 0; neighbor < graph[node].size(); ++neighbor) {
        if (graph[node][neighbor]) {
            if (!visited[neighbor] && dfsCycle(neighbor, graph, visited, recStack, path))
                return true;
            else if (recStack[neighbor]) {
                path.push_back(neighbor);
                return true;
            }
        }
    }

    path.pop_back();
    recStack[node] = false;
    return false;
}

int main() {
    int n;
    cout << "Enter number of vertices: ";
    cin >> n;

    vector<vector<int>> graph(n, vector<int>(n));
    cout << "Enter adjacency matrix:\n";
    for (int i = 0; i < n; ++i)
        for (int j = 0; j < n; ++j)
            cin >> graph[i][j];

    vector<bool> visited(n, false);
    vector<bool> recStack(n, false);
    vector<int> path;

    for (int i = 0; i < n; ++i) {
        if (!visited[i] && dfsCycle(i, graph, visited, recStack, path)) {
            cout << "Cycle found: ";
            for (int v : path) cout << v << " ";
            cout << endl;
            return 0;
        }
    }

    cout << "No cycle found.\n";
    return 0;
}

File: CyclicDFS.java
import java.util.*;

public class CyclicDFS {
    static boolean dfs(int node, boolean[] visited, boolean[] recStack, List<List<Integer>> graph, List<Integer> path) {
        visited[node] = true;
        recStack[node] = true;
        path.add(node);

        for (int neighbor : graph.get(node)) {
            if (!visited[neighbor] && dfs(neighbor, visited, recStack, graph, path))
                return true;
            else if (recStack[neighbor]) {
                path.add(neighbor); // for printing the cycle
                return true;
            }
        }

        recStack[node] = false;
        path.remove(path.size() - 1);
        return false;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number of vertices: ");
        int n = sc.nextInt();

        List<List<Integer>> graph = new ArrayList<>();
        for (int i = 0; i < n; i++) graph.add(new ArrayList<>());

        System.out.println("Enter adjacency matrix:");
        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (sc.nextInt() == 1)
                    graph.get(i).add(j);

        boolean[] visited = new boolean[n];
        boolean[] recStack = new boolean[n];
        List<Integer> path = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            if (!visited[i] && dfs(i, visited, recStack, graph, path)) {
                System.out.println("Cycle detected: " + path);
                return;
            }
        }

        System.out.println("No cycle found.");
    }
}
